### PR TITLE
New feature: Auto reruns minimum duration

### PR DESCRIPTION
**Pull request by nhoobin**
This builds upon an earlier branch for the workflow export history.

### CODE REVIEW

File: classes/output/workflowhistory/renderer.php [hunk 1]
```diff
@@ -86,8 +86,14 @@ public function render_workflowhistory_table($workflowid, $searchparams = [], $d
...
         $namefields = get_all_user_name_fields(true, 'u');
         $sqlfields = "tfh.*, {$namefields}";
-        $sqlfrom = '{tool_trigger_workflow_hist} tfh LEFT JOIN {user} u ON tfh.userid = u.id';
-        $sqlwhere = 'workflowid = :workflow';
+        $sqlfrom = '{tool_trigger_workflow_hist} tfh';
+
+        $joins = [
+            'LEFT JOIN {user} u ON tfh.userid = u.id',
+        ];
+        $sqlfrom = $sqlfrom . ' ' . implode(' ', $joins);
```
Comment (Peterburnett) [inline classes/output/workflowhistory/renderer.php:94]: This seems overly complicated for the one line. Is there reason to make this extendable currently?

File: db/upgrade.php [hunk 2]
```diff
@@ -332,5 +332,28 @@ function xmldb_tool_trigger_upgrade($oldversion) {
...
+        $table = new xmldb_table('tool_trigger_workflow_hist');
+        $field = new xmldb_field('attemptnum', XMLDB_TYPE_INTEGER, '10', null, null, null, '1', 'errorstep');
+
+        // Conditionally launch add field id.
+        if (!$dbman->field_exists($table, $field)) {
+            $dbman->add_field($table, $field);
+        }
+
+        // Update the new attemptnum field with the default number of attempts.
+        $rs = $DB->get_recordset('tool_trigger_workflow_hist');
```
Comment (Peterburnett) [inline db/upgrade.php:347]: I think you get this for free if its not nullable? Could be wrong

File: lang/en/tool_trigger.php [hunk 3]
```diff
@@ -94,12 +96,23 @@
...
 $string['errorsavingworkflow'] = 'Something went wrong while attempting to save the workflow. Please try again.';
 $string['errorstep'] = 'Errored step {$a}';
+$string['errorstepretrypending'] = '(Pending retry {$a})';
 $string['executenext'] = 'Execute following step';
 $string['executenextconfirm'] = 'Are you sure you wish to execute following step? It will execute based on the data from this step.';
 $string['expectedresponse'] = 'Expected response code';
 $string['expectedresponse_help'] = 'Enter the HTTP Response code you expect to receive from the step. Any other response header will cause the step to error.';
 $string['failedstep'] = 'Failed step {$a}';
 $string['filter'] = 'Filter';
+$string['filtercancelled'] = 'Cancelled';
```
Comment (Peterburnett) [inline lang/en/tool_trigger.php:106]: Most of these strings might be in core?

File: classes/output/workflowhistory/renderer.php (hunk 1 above)
Comment (nhoobin) [reply classes/output/workflowhistory/renderer.php:94]: Ah, that was leftover from when I was joining additional tables. I can remove this.

**Pull request merged**